packages/vite/src/node/packages.ts

```diff
@@ -201,31 +207,35 @@ export function loadPackageData(pkgPath: string): PackageData {
     hasSideEffects = () => null
   }
 
+  const resolvedCache: Record<string, string | undefined> = {}
   const pkg: PackageData = {
     dir: pkgDir,
     data,
     hasSideEffects,
-    webResolvedImports: {},
-    nodeResolvedImports: {},
-    setResolvedCache(key: string, entry: string, targetWeb: boolean) {
-      if (targetWeb) {
-        pkg.webResolvedImports[key] = entry
-      } else {
-        pkg.nodeResolvedImports[key] = entry
-      }
+    setResolvedCache(key, entry, options) {
+      resolvedCache[getResolveCacheKey(key, options)] = entry
     },
-    getResolvedCache(key: string, targetWeb: boolean) {
-      if (targetWeb) {
-        return pkg.webResolvedImports[key]
-      } else {
-        return pkg.nodeResolvedImports[key]
-      }
+    getResolvedCache(key, options) {
+      return resolvedCache[getResolveCacheKey(key, options)]
     },
   }
 
   return pkg
 }
 
+function getResolveCacheKey(key: string, options: InternalResolveOptions) {
+  // cache key needs to include options which affect
+  // `resolvePackageEntry` or `resolveDeepImport`
+  return [
+    key,
+    options.webCompatible ? '1' : '0',
+    options.isRequire ? '1' : '0',
+    options.conditions.join('_'),
+    options.extensions.join('_'),
+    options.mainFields.join('_'),
+  ].join('|')
+}
```

Comment:
It is a pity that we couldn't use the environment instance as the key here. About the perf impact, did you measure the time to generate this key? Maybe we could avoid the array plus join and directly concatenate here.

Another option we could explore (maybe later?) is to have the `resolveCache` be created for each environment and pass that as an object in `options`, so we can directly do `options.resolveCache.get(key)` and avoid the key.

Reply:
> About the perf impact, did you measure the time to generate this key? Maybe we could avoid the array plus join and directly concatenate here.

Yeah, I would imagine `getResolveCacheKey` is the factor. From the flamegraph, there is a stack of `resolvePackageEntry -> getResolvedCache -> getResolveCacheKey` and `getResolveCacheKey` is what's costly.

![image](https://github.com/user-attachments/assets/fcf17f2e-4871-4b0c-be4a-ad1fa8b8eaf8)

Reply:
> Maybe we could avoid the array plus join and directly concatenate here.

I tried to do some random tweak like this, but I didn't see much difference.

```ts
return (key 
  + '|' + (options.webCompatible ? '1' : '0')
  + '|' + (options.isRequire ? '1' : '0')
  + '|' + (options.conditions[0])
  + '|' + (options.conditions[1])
  + '|' + (options.conditions[1])
  + '|' + (options.extensions[2])
  + '|' + (options.extensions[2])
  // ...
)
```

Reply:
To avoid concatenating `conditions, extensions, mainFields`, we can keep track of yet another object map like this:

```ts
const objectKeyMap = new WeakMap<object, string>();

function getOptionObjectKey(key: object) {
  let value = objectKeyMap.get(key);
  if (!value) {
    value = Math.random().toString(36).slice(2);
    objectKeyMap.set(key, value);
  }
  return value;
}

function getResolveCacheKey(key: string, options: InternalResolveOptions) {
    return (
      key
      + '|' + (options.webCompatible ? '1' : '0')
      + '|' + (options.isRequire ? '1' : '0')
      + '|' + getOptionObjectKey(options.conditions)
      + '|' + getOptionObjectKey(options.extensions)
      + '|' + getOptionObjectKey(options.mainFields)
    )
}
```

This seems to help a lot though I'm not sure how much of optimization is the goal since this is still a fraction of the entire `resolveId` 80000 times.

![image](https://github.com/user-attachments/assets/9d3f78b0-2320-4822-a24e-0de24a2c8230)